Add events for player changing spectator target

CalebBassham opened the pull request:
- Add PlayerStartSpectatingEntityEvent
- Add PlayerStopSpectatingEntityEvent

This is my first pr here so if there is anything that should be different please let me know!

CalebBassham committed:
Add events for player changing spectator target

- Add PlayerStartSpectatingEntityEvent
- Add PlayerStopSpectatingEntityEvent

CalebBassham committed:
Fixed some issues with pr

- Added some docs for new events
- Readded unused imports that intellij removed automatically
- Changed how events are called

CalebBassham committed:
Made some requested revisions

- Removed blank line from import statements
- Removed my events from import statements
- reverted the name of a variable to reduce diffs

zachbr commented on Spigot-API-Patches/0157-Add-spectator-target-events.patch:
```diff
@@ -0,0 +1,121 @@
++        this.newSpectatorTarget = newSpectatorTarget;
++    }
++
++    public Entity getCurrentSpectatorTarget() {
```
Even basic docs like "The entity being spectated" often lend clarity, please add them API side.

zachbr commented on Spigot-API-Patches/0157-Add-spectator-target-events.patch:
```diff
@@ -0,0 +1,121 @@
++        return currentSpectatorTarget;
++    }
++
++    public Entity getNewSpectatorTarget() {
```
Docs

zachbr commented on Spigot-API-Patches/0157-Add-spectator-target-events.patch:
```diff
@@ -0,0 +1,121 @@
++        this.spectatorTarget = spectatorTarget;
++    }
++
++    public Entity getSpectatorTarget() {
```
Docs

zachbr commented on Spigot-Server-Patches/0381-Call-player-spectator-target-events.patch:
```diff
@@ -0,0 +1,51 @@
+-        this.cv = (Entity) (entity == null ? this : entity);
+-        if (entity1 != this.cv) {
++        if (newSpectatorTarget == null) {
++            this.cv = this;
```
all of these assignments to `this.cv` should probably use an obfhelper setter, with that setter up by the field declaration.
```java
private Entity cv; private void setSpectatorTarget(Entity e) { this.cv = e; } // Paper - OBFHELPER
```

Just so when we go to update this to 1.14 or 1.5512.4 or whatever its clear what the intent was even when this gets renamed.

CalebBassham replied:
`setSpectatorTarget(Entity)` is the method that I am changing for this pr. Should I do something like `setSpectatorTargetField(Entity)` as an obfhelper or keep it as it is?

zachbr replied:
Naming it slightly different is probably smart, I think that'd be fine 👍

CalebBassham committed:
Add more docs and add OBFHELPER

Player{Start,Stop}SpectatingEntityEvent Docs
OBFHELPER for setting spectator target

CalebBassham committed:
Remove accidental import

zachbr committed:
Merge patches, minor doc changes

zachbr committed:
Minor changes to code

getBukkitEntity() is usually preferred over a direct cast

Shouldn't need to double-set the field if new spectator is null, it
should be set at the end regardless

zachbr committed:
Rebuild patches for rebase

zachbr committed:
Fix typo in doc change
